src/core/SampleBufferCache.cpp

```diff
@@ -0,0 +1,80 @@
+/*
+ * SampleBufferCache.cpp - Cache for storing SampleBuffer's
+ *
+ * Copyright (c) 2024 saker <sakertooth@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SampleBufferCache.h"
+
+#include <cassert>
+#include <cerrno>
+#include <stdexcept>
+
+namespace lmms {
+
+SampleBufferCache::SampleBufferCache(size_t capacity)
+	: m_capacity(capacity)
+{
+	if (capacity == 0) { throw std::runtime_error{"Capacity of cache cannot be zero"}; }
+}
+
+auto SampleBufferCache::add(const QString& key, Buffer buffer) -> bool
+{
+	if (buffer == nullptr || m_cache.contains(key)) { return false; }
+
+	assert(m_cache.size() <= m_capacity);
+	if (m_cache.size() == m_capacity)
+	{
+		const auto leastRecentlyUsed = std::min_element(m_cache.begin(), m_cache.end(),
+			[](const auto& entryOne, const auto& entryTwo) { return entryOne.use_count() < entryTwo.use_count(); });
+		m_cache.erase(leastRecentlyUsed);
+	}
+
+	m_cache.insert(key, buffer);
```

Comment:
Could also add the path to the file system watcher here.

Reply:
I don't want to couple the watcher to the cache.

Reply:
Why not? The watcher is needed for the cache to function properly and you shouldn't have to remember to add the watcher in calling code.

Reply:
SRP, but I think it's okay. I was thinking I couldn't test the cache if I coupled the watcher to it but I can still test without it.